### PR TITLE
[WIP][Example] GraphSAGE model with 0.5 sampler and multi-GPU implementation

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
A working end-to-end example of using the 0.5 sampler interface in a multi-GPU setting.

This is based on and compared against #1143 .

Several notes:

1. Worker processes should be created with `mp.Process` to make sure that the graph structure and data is not duplicated.  This leads to the following issue:
2. Neighbor samplers in worker processes freezes because of the long-standing deadlock bug in GNU OpenMP with `fork` (see #1129 ).  https://github.com/pytorch/pytorch/issues/17199 proposed a workaround and I adopted it in this example.  However, we need to confirm whether this also applies for MXNet and Tensorflow training (**help wanted**).

### Performance comparison

Here is a table computing the throughput of training examples processed per second.  The testbed is a p3.8xlarge AWS instance and the NodeFlow implementation is taken from #1143.

| Implementation | Average epoch time | Training example processed per second | Accuracy |
|:----:|:----:|:----:|:----:|
| NodeFlow | 33.72 | 6508 | 0.9553 |
| Current | 19.97 | 9067 | 0.9523 |
| NodeFlow, 4 GPUs | 9.093 | 24473 | 0.9604 |
| Current, 4 GPUs | 5.439 | 28342 | 0.9538 |

I believe the NodeFlow implementation can be optimized further, e.g. removing additional overheads in repeated computation of `IsMultigraph()` (@zheng-da )

### TODOs

- [ ] Unsupervised node classification example
- [ ] (Optional) Test on a dataset larger than Reddit (preferably one in Open Graph Benchmark, @mufeili any suggestions?)
- [ ] Tune the accuracy.  NodeFlow implementation and current implementation actually sample a little bit differently (for instance, the seed nodes has their neighbors resampled in the first layer in the current implementation, while the NodeFlow implementation reuses the ones in the second layer), and I wonder if this is the reason I had a minor performance drop.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by BarclayII** [comment]
Closing this since I'm basing on a wrong branch.
